[GSoC] Add EUF Theory Solver and DPLL(T) Integration with Public API euf_ask

krishnavbajoria02 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

***

**Description**  

- **Implements core EUF theory solver**
  - Adds `EUFCongruenceClosure` for congruence closure on equalities and disequalities.
  - Adds `EUFTheorySolver` class with:
    - `from_encoded_cnf`
    - `assert_lit`
    - `Backtrack`
    - `IsTrue`
    - `Explanation`
  - Maintains two‑way mapping between encoded integer literals and SymPy expressions for SAT ↔ theory communication.

- **Adds public EUF API**
  - Introduces new `euf_ask` function in `sympy/assumptions/euf_ask.py`.
  - Implements `_preprocess_euf` to normalise formulas so only `Q.eq` / `Q.ne` predicates remain at the top level.
  - Applies currying and dummy symbol assignment to handle arbitrary function arguments safely.

- **Integrates into DPLL(T) flow**
  - Updated `logic/inference.satisfiable` and `dpll_satisfiable` in `dpll2.py` to support `use_euf_theory`.
  - `SATSolver` now calls `EUFTheorySolver` during model search, similar to the integration for `LRASolver`.

- **Adds comprehensive tests**
  - Unit tests for `EUFTheorySolver` covering: initialization, literal assertion, conflict detection, transitivity, trivial conflicts, AppliedPredicate handling.
  - Integration tests simulating SAT‑theory interaction by asserting integer IDs from encoded CNFs and calling `check()`.
  - API‑level tests for `euf_ask` to validate end‑to‑end functionality.

- **Current status**
  - Basic equality/disequality reasoning, transitive closure, and conflict detection are working.
  - Predicate implication graph, optimised backtracking, and extended predicate coverage are planned for future work.



#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

NO ENTRY

<!-- END RELEASE NOTES -->

krishnavbajoria02 committed:
Add EUF Congruence Closure Engine and Symbolic Tests for DPLL(T) Integration

krishnavbajoria02 committed:
Remove trailing whitespace

krishnavbajoria02 committed:
Remove trailing whitespace

krishnavbajoria02 committed:
Correcting type errors

krishnavbajoria02 committed:
feat(logic/euf_theory): use SymPy symbols and custom EUF function applications directly

krishnavbajoria02 committed:
Removed trailing whitespace

krishnavbajoria02 committed:
Removed unicode characters

krishnavbajoria02 committed:
Removed unicode characters

krishnavbajoria02 committed:
Added import for symbols

krishnavbajoria02 committed:
Remove unwanted doctests

krishnavbajoria02 committed:
Remove doctests

krishnavbajoria02 committed:
test(core): add test for euf_theory.Apply in test_args.py to satisfy Basic subclass coverage

krishnavbajoria02 committed:
Added new line at the end of the file

krishnavbajoria02 committed:
Remove unicode character

krishnavbajoria02 committed:
Remove unwanted comment

krishnavbajoria02 committed:
Remove unwanted space

krishnavbajoria02 committed:
Added curry function in Lambda class and added tests for it

krishnavbajoria02 committed:
Updated docstring and corrected doctests

krishnavbajoria02 committed:
Removed EUFFunction and Apply and used Lambda objects

krishnavbajoria02 committed:
Removed unused variables

krishnavbajoria02 committed:
Commented out tests temporarily

krishnavbajoria02 committed:
Refactored code and added tests

krishnavbajoria02 committed:
Added new line at the end of the file  and modified tests for currification

krishnavbajoria02 committed:
Remove unwanted code

krishnavbajoria02 committed:
Added conditions for fixing failing test cases

krishnavbajoria02 committed:
Removed trailing whitespace

krishnavbajoria02 committed:
Added new line at the end of the file

krishnavbajoria02 committed:
Removed unicode characters

krishnavbajoria02 committed:
Updated documentation and variable names matching the paper

krishnavbajoria02 committed:
Removed unicode character

krishnavbajoria02 committed:
Remove unwanted variables

krishnavbajoria02 committed:
Rewritten using set comprehension

krishnavbajoria02 committed:
Remove unwanted import

krishnavbajoria02 committed:
Corrected doctests by adding necessary imports

krishnavbajoria02 committed:
Empty commit

krishnavbajoria02 committed:
WIP: Add EUF Theory Solver and DPLL(T) Integration with Public API euf_ask

krishnavbajoria02 committed:
Corrected the union function in EUFCongruenceClosure

krishnavbajoria02 committed:
Corrected tests and refactored code for preprocessing

krishnavbajoria02 committed:
Implemented the EUFTheorySolver and refactored some code for integration purposes

krishnavbajoria02 committed:
Remove unnecessary variables and imports

krishnavbajoria02 commented:
I also ran the doctests and I think they work fine and they should pass but they don't and are getting cancelled here. I have no idea as to why is that happening.

krishnavbajoria02 committed:
Refactored code for failing tests

asmeurer commented:
Why don't the tests on CI show which test files are being run? @oscarbenjamin

asmeurer commented:
If you look at the pyodide test failures there are some failures there that give actual tests that are failing.

oscarbenjamin commented:
That's just the default output of `pytest -n auto` which runs the tests on multiple cores. Maybe it should use `-v` or `-vv`.

krishnavbajoria02 committed:
Corrected conditions in find_model and corrected type of Exception

krishnavbajoria02 committed:
Corrected immediate_conflicts which was being overwritten instead of accumulated

krishnavbajoria02 committed:
Remove redundant test

krishnavbajoria02 commented:
@TiloRC @asmeurer can you please review this PR now. All tests pass. I'll soon add more documentation and add tests for checking against z3.

krishnavbajoria02 committed:
Added documentation and doctests

krishnavbajoria02 committed:
Remove unwanted import

krishnavbajoria02 committed:
Added necessary imports in doctests

krishnavbajoria02 committed:
Refactored code and added ProofProducingCongruenceClosure

asmeurer commented on sympy/logic/algorithms/euf_theory_solver.py:
```diff
@@ -209,151 +165,119 @@ def process_pred(pred):
+            return consequences
+
         except ValueError as e:
             if "EUF conflict" in str(e):
```
It's better to just use a custom exception for this and catch that specific exception.

krishnavbajoria02 replied:
I have made the changes and added two new exception classes. Can you please review it now? Thanks.

TiloRC commented on sympy/logic/algorithms/euf_theory_solver.py:
```diff
@@ -0,0 +1,472 @@
+            rep_lhs = self.cc._find(lhs_c)
+            rep_rhs = self.cc._find(rhs_c)
+
+            # Check for immediate disequality conflict
+            for (da, db) in self.disequality_causes.keys():
+                if ((self.cc._find(da) == rep_lhs and self.cc._find(db) == rep_rhs) or
+                    (self.cc._find(da) == rep_rhs and self.cc._find(db) == rep_lhs)):
+                    raise ValueError("EUF conflict: equality contradicts disequality")
+
+            # Add equality with reason for proof production
+            self.cc.add_equality_with_reason(lhs, rhs, ceq)
```
I believe this block of code is responsible for the bug. The value error is raised before ` self.cc.add_equality_with_reason`  can be called. As a consequence, when you're creating explanations, `cc` is not aware of this equality.

krishnavbajoria02 replied:
I hope this query is also now resolved after our last meet conversation.

krishnavbajoria02 committed:
Corrected conditions and code and added tests

krishnavbajoria02 committed:
Removed implicit import

krishnavbajoria02 committed:
Removed unnecessary imports and unused variables

krishnavbajoria02 committed:
Remoe unused variable

krishnavbajoria02 commented:
@asmeurer @TiloRC 
Could you please review the changes here. I have also added some tests that are failing. After running the debugger what I am able to find is that majorly the problem just lies in ```explain_equality``` function. I am currently working on that.

krishnavbajoria02 committed:
Added more tests in test_euf_solver.py

krishnavbajoria02 committed:
Update exceptions being handled

krishnavbajoria02 committed:
Added more conditions and tests along with xfail tests

krishnavbajoria02 committed:
Added newline at the end of the file

krishnavbajoria02 committed:
Remove unwanted tests and variables

krishnavbajoria02 committed:
Removed XFAIL decorator

krishnavbajoria02 committed:
Removed unnecessary import

krishnavbajoria02 committed:
Corrected conditions and tests

TiloRC commented on sympy/logic/algorithms/dpll2.py:
```diff
@@ -234,7 +245,8 @@ def _find_model(self):
                     else:
                         res = None
                     if res is None or res[0]:
-                        yield {self.symbols[abs(lit) - 1]:
+                        if not self.euf:
+                            yield {self.symbols[abs(lit) - 1]:
                                     lit > 0 for lit in self.var_settings}
                     else:
```
this looks suspicious. If euf is being used how do the results get yielded?

krishnavbajoria02 replied:
There is another condition check for `if self.euf` later on.

TiloRC commented on sympy/logic/algorithms/dpll2.py:
```diff
@@ -46,12 +47,21 @@ def dpll_satisfiable(expr, all_models=False, use_lra_theory=False):
-    solver = SATSolver(expr.data + immediate_conflicts, expr.variables, set(), expr.symbols, lra_theory=lra)
+
+    if use_euf_theory:
+        euf, euf_conflicts = EUFTheorySolver.from_encoded_cnf(expr)
+        immediate_conflicts.extend(euf_conflicts)
```
Maybe try not using the immediate conflicts here? It's possible this could be causing a bug and it's not necessary.

TiloRC commented on sympy/logic/algorithms/euf_theory_solver.py:
```diff
@@ -0,0 +1,720 @@
+            if is_positive:
+                consequences = self.SetTrue(constraint)
+            else:
+                if isinstance(constraint, Eq):
+                    new_eq = Ne(constraint.lhs, constraint.rhs)
+                    consequences = self.SetTrue(new_eq)
+                else:  # Ne
+                    new_eq = Eq(constraint.lhs, constraint.rhs)
+                    consequences = self.SetTrue(new_eq)
```
I would be good to be able to assume everything is an Eq by preprocessing things earlier so we could simplify this logic.

TiloRC commented on sympy/logic/algorithms/euf_theory_solver.py:
```diff
@@ -0,0 +1,720 @@
+
+        try:
+            if is_positive:
+                consequences = self.SetTrue(constraint)
```
Don't you need to check if it's a disequality here?

TiloRC commented on sympy/logic/algorithms/euf_theory_solver.py:
```diff
@@ -0,0 +1,720 @@
+        return literal.lhs, literal.rhs, True
+    elif isinstance(literal, Ne):
+        return literal.lhs, literal.rhs, False
+    elif isinstance(literal, Not) and isinstance(literal.args[0], Eq):
+        eq = literal.args[0]
+        return eq.lhs, eq.rhs, False
```
Is it possible for this condition to be true? Due to sympy's automatic simplification I don't think it should be possible.

krishnavbajoria02 replied:
Yes, this is another instance of redundant code. While writing I was just being careful to not miss any of such conditions even if that meant redundancy at that point due to my main focus at the time being correctness then.

TiloRC commented on sympy/logic/algorithms/euf_theory_solver.py:
```diff
@@ -0,0 +1,720 @@
+
+        return conflict_clause
+
+    def SetTrue(self, literal):
```
I might be a bit better for this to take three arguments: lhs, rhs, is_positive. I think that might simplify the logic a small amount.

krishnavbajoria02 replied:
I hope this too is resolved based off our last meet conversation.

TiloRC commented on sympy/logic/algorithms/euf_theory_solver.py:
```diff
@@ -0,0 +1,720 @@
+        self.timestamp_counter += 1
+        lhs, rhs, is_positive = _canonical_lit(literal)
+        ceq = _canon_eq(lhs, rhs)
+        self.literal_terms.setdefault(ceq, (lhs, rhs, is_positive))
```
Not sure if this matters at all but maybe lhs and rhs being in different order than ceq might have an impact on something.

TiloRC commented on sympy/logic/algorithms/euf_theory_solver.py:
```diff
@@ -0,0 +1,720 @@
+            rep_lhs = self.cc._find(lhs_c)
+            rep_rhs = self.cc._find(rhs_c)
+
+            # Check for immediate disequality conflict BEFORE adding equality
+            for diseq_pair, diseq_cause in self.disequality_causes.items():
+                da, db = diseq_pair
+                if ((self.cc._find(da) == rep_lhs and self.cc._find(db) == rep_rhs) or
+                    (self.cc._find(da) == rep_rhs and self.cc._find(db) == rep_lhs)):
+                    # Found conflict - equality contradicts disequality
+                    self.current_conflict_diseq = diseq_cause
+                    self.current_conflict_eq = ceq
+                    self._current_exception_type = 'disequality_contradiction'
+                    raise EUFDisequalityContradictionException("EUF conflict: Equality contradicts Disequality")
+
+            # Add equality with reason for proof production
+            if rep_lhs != rep_rhs:
+                self.cc.add_equality_with_reason(lhs, rhs, ceq)
```
Didn't we say we should add the equality before raising the exception a week ago?

TiloRC commented on sympy/logic/algorithms/euf_theory_solver.py:
```diff
@@ -0,0 +1,720 @@
+            rep_lhs = self.cc._find(lhs_c)
+            rep_rhs = self.cc._find(rhs_c)
+
+            # Check for immediate equality conflict
+            if rep_lhs == rep_rhs:
+                # Terms are already equal - disequality contradicts equality
+                self.current_conflict_diseq = ceq  # The disequality we're trying to assert
+                self.current_conflict_eq = None
+                self._current_exception_type = 'equality_contradiction'
+                raise EUFEqualityContradictionException("EUF conflict: disequality within same equivalence class")
+
+            lhs, rhs, is_positive = _canonical_lit(literal)
+            if self.cc.are_equal(lhs,rhs):
+                # Terms are already equal - disequality contradicts equality
+                self.current_conflict_diseq = ceq  # The disequality we're trying to assert
+                self.current_conflict_eq = None
+                self._current_exception_type = 'equality_contradiction'
+                raise EUFEqualityContradictionException("EUF conflict: disequality within same equivalence class")
```
There seems to be a lot of duplicated lines of code here. Could this be simplified?

krishnavbajoria02 replied:
Right there are these cases where I have duplicated certain code. I have removed the redundant code now in my recent changes. Please do review them. Thanks.

krishnavbajoria02 committed:
Refactored code and modified algorithm for explain_equality

krishnavbajoria02 committed:
Added new line at the end of file

krishnavbajoria02 committed:
Refactor reset function

krishnavbajoria02 committed:
Added documentation and corrected further conditions

krishnavbajoria02 commented on sympy/logic/tests/test_euf_solver.py:
```diff
@@ -0,0 +1,641 @@
+    assert isinstance(explanation, set)
+
+
+def test_reset_functionality():
+    """Test reset method."""
+    solver = EUFTheorySolver()
+    solver.Initialize({Eq(a, b), Ne(c, d)})
+
+    solver.SetTrue(Eq(a, b))
+    solver.SetTrue(Ne(c, d))
+
+    # Reset should clear all state
+    solver.reset()
+
+    assert len(solver.interpretation_stack) == 0
+    assert solver.timestamp_counter == 0
+    assert len(solver.literal_timestamp) == 0
+    assert len(solver.disequalities_set) == 0
+    assert len(solver.disequality_causes) == 0
```
Right. After our meet I was thinking about removing this entire reset function for this PR atleast since we are not that keen right now to add an incremental based solver so just creating a new solver seems fine to me here.

krishnavbajoria02 committed:
Removed redundant conditions

krishnavbajoria02 commented on sympy/logic/algorithms/euf_theory_solver.py:
```diff
@@ -0,0 +1,472 @@
+
+    def SetTrue(self, literal):
+        """
+        Assert literal and return T-consequences.
```
I hope this also is now resolved after our last conversation in the meet.

krishnavbajoria02 commented on sympy/logic/algorithms/euf_theory_solver.py:
```diff
@@ -209,151 +165,119 @@ def process_pred(pred):
         constraints = self.literal_eqs[abs_enc]
+
         try:
+            consequences = set()
```
Yeah I think now that per literal since there is just only one equality/disequality we can get rid of the list here.

krishnavbajoria02 committed:
Modified changes to ask and added disequality explanations and added tests for it

krishnavbajoria02 committed:
Remove wrong test

krishnavbajoria02 committed:
Removed majority of XFAIL tests that now pass

krishnavbajoria02 committed:
Added tests related to uninterpreted functional applications

krishnavbajoria02 committed:
Added random test generator to check against z3

krishnavbajoria02 committed:
Added functional equality/disequality explanation and added tests and integrated conflict based clause learning

krishnavbajoria02 committed:
Removed unicode characters

krishnavbajoria02 commented:
> After it adds that conflict clause the sat solver will find that the whole query is unsatisfiable.

You mean satisfiable at the end right, since the `can_be_true` will be `True` and `can_be_false` will be `False`.

krishnavbajoria02 committed:
Removed unwanted functions and added diamond disequality tests

krishnavbajoria02 committed:
empty commit

krishnavbajoria02 commented:
> It would be good to add tests for the diamond disequality https://www.imn.htwk-leipzig.de/~waldmann/edu/ss12/cp/folien/main/node127.html

I have added tests for the diamond disequality. Please do review them. Thanks.

krishnavbajoria02 committed:
Remove unwanted import

krishnavbajoria02 committed:
Integrated calls to euf theory solver from ask using satask with additional parameter and also added documentation to euf theory solver

krishnavbajoria02 committed:
Remove unwanted import

krishnavbajoria02 committed:
Corrected import of z3

krishnavbajoria02 committed:
Removed unwanted examples

TiloRC committed:
Merge branch 'master' into krishnavbajoria02-euf-integration

TiloRC committed:
logic: Fix and improve euf random tests

- fixed bug where functions did not have consistent arities
- started using z3 wrapper instead of Z3Comparator to
  determine satisfiablity of random problems
- ensured that RandomEUFTestGenerator always generates the correct
  number of equalities
- ensured there's a balance of satisfiable and unsatisfiable problems
- added test involving large sat problem

TiloRC commented:
Also, it would probably be a good idea to move all of the random tests to the end of the file as if they fail it prevents the other tests from running.

krishnavbajoria02 committed:
Added more functionality to deal with expressions in congruence closure

krishnavbajoria02 committed:
Moidified conditions and parameters

krishnavbajoria02 committed:
Removed unnecessary imports

krishnavbajoria02 committed:
Added fallback/skip for failed z3 installation

krishnavbajoria02 commented:
> @krishnavbajoria02 I made some improvements to the random problem generation and was able to make a test, `test_random_sat_problems`, which sometimes fails due to what is likely a bug with euf -- z3 and sympy don't agree on the satisfiability of some of the problems it generates.

What I observe is that when the number of trials are just 1 then it doesnt throw out an assertion error on all the 5-6 times I tried but the moment we increase it to 2 then it fails. This made me question if it somthing related to the recursion depth issue that we were talking about at the pretty start when we I was writing the EUFCongruenceClosure system since there also once the number of terms increased beyond 200 as it does here then some errors had started to come. But I am not sure if they are related or not. It also makes it a whole lot harder to debug as such given the sheer size of the input constraints list.

krishnavbajoria02 committed:
Updated test

krishnavbajoria02 commented:
So I spent a lot of time debugging the `test_random_sat_problems` and I found that if i pass the 'AND' expressions individually to the satisfiable and finally then take the 'OR' of it, then it returns the correct result. For some reason I think the OR of AND is not being properly converted to its proper `CNF`, because otherwise the correct result only is being determined. Cause if OR of AND not being properly read as it is then what is happening is that equalities/disequalities in one AND expression is merging or leaking into another AND expression causing it to return False. This could also be a problem in the way the sat solver is interacting with the euf solver. It was just my observation that I noticed that if I divide it into a two step process it returns the correct answer.

krishnavbajoria02 commented:
<img width="1329" height="849" alt="Screenshot (459)" src="https://github.com/user-attachments/assets/7e64fc76-c556-48a9-b13e-04c1f58ab603" />
Also these errors seem unrelated.

krishnavbajoria02 committed:
Added additional random tests

TiloRC commented:
ask(Q.real(x), Q.eq(x, y) & Q.integer(y))

To handle this example, the sat solver needs to find that ~Q.real(x) & Q.eq(x, y) & Q.integer(y) is unsatisfiable. With the EUF theory solver and the rules clauses it should be able to handle this.

One of the rules clauses that would be added would be the fact  Q.integer(y) implies Q.real(y). This fact can be represented with the clause ~Q.integer(y) | Q.real(y) -- if Q.integer(y) is true the only way for the whole clause to be true is that Q.real(y) be true.  So after we add the rules clauses we will get a formula that looks like this:

~Q.real(x) & Q.eq(x, y) & Q.integer(y) & (~Q.integer(y) | Q.real(y) ) & (more rules clauses)

We can ignore the more rules clauses for the sake of simplicity.  Since Q.integer(y) must be true, there is only one possible satisfying assignment to 

~Q.real(x) & Q.eq(x, y) & Q.integer(y) & (~Q.integer(y) | Q.real(y) ) 

That satisfying assignment is:
Q.real(x) : false 
Q.eq(x, y) : true 
Q.integer(y) : true 
 Q.real(y) : true 

However, the euf solver will find that  Q.real(x), Q.eq(x, y),  Q.real(y) conflict with each other.

TiloRC commented on sympy/logic/algorithms/euf_theory_solver.py:
```diff
@@ -0,0 +1,944 @@
+        encoded_items = (sorted(encoded_cnf.encoding.items(), key=lambda x: str(x[0]))
+                        if testing_mode else encoded_cnf.encoding.items())
+
+        for pred, enc in encoded_items:
+            try:
+                eqs, triv = process_pred(pred)
+
+                if eqs:
+                    literal_eqs[enc] = eqs
+                    solver._enc_to_lit[enc] = eqs[0]
+                    solver.lit_to_enc[eqs[0]] = enc
+                    if isinstance(eqs[0], Eq):
+                        opp_equation = Ne(eqs[0].lhs, eqs[0].rhs)
+                        solver._enc_to_lit[-enc] = opp_equation
+                        solver.lit_to_enc[opp_equation] = -enc
+                    if isinstance(eqs[0], Ne):
+                        opp_equation = Eq(eqs[0].lhs, eqs[0].rhs)
+                        solver._enc_to_lit[-enc] = opp_equation
+                        solver.lit_to_enc[opp_equation] = -enc
+
+                # Handle trivial conflicts
+                if True in triv:
+                    conflicts.append([enc])
+                if False in triv:
+                    conflicts.append([-enc])
+
+            except EUFUnhandledInput:
+                continue
```
This is something that should probably be dealt with at the level of ask and not the euf solver.

krishnavbajoria02 replied:
> Maybe this is something that should probably be dealt with at the level of ask (or maybe the sat solver?) and not the euf solver. Though the euf solver should raise an exception maybe if the encoding is bad.

No actually this is my mistake since I have added both encodings like if I just have in my set of equations the term Eq(a,b) that has encoding 1 in my euf_theory_solver I do the encoding of it as 1 and then create an additional encoding of Ne(a,b) as -1, maybe I should not do that.

TiloRC replied:
Keep in mind that when you created explanations you'll need to know which literals in the sat solver are associated with which of these equalities and inequalities. 

Make sure to add tests for when the encoded cnf has distinct atoms for logically equivalent equalities or disequalities.

krishnavbajoria02 replied:
I have fixed the problems with the encoding in my current commit. Can you please review it now. Thanks.

TiloRC commented on sympy/logic/algorithms/euf_theory_solver.py:
```diff
@@ -0,0 +1,944 @@
+                    left, right = pred.lhs, pred.rhs
+                    is_eq = pred.func == Eq
+
+                # Trivial cases
+                if left == right:
+                    if is_eq:
+                        return [], [True]
+                    else:
+                        return [], [False]
```
Do we need this?

krishnavbajoria02 replied:
Like this is for immediate conflict. So if both lhs and rhs are equal so trivially we add them as True otherwise false if not is_eq.

krishnavbajoria02 replied:
I have removed these two lines of code in my current commit.

krishnavbajoria02 commented on sympy/logic/algorithms/euf_theory_solver.py:
```diff
@@ -0,0 +1,944 @@
+                    else:
+                        return [], [False]
+
+                # Create constraint directly
+                if is_eq:
+                    new_constraint = Eq(left, right)
+                else:
+                    new_constraint = Ne(left, right)
```
Ok yeah this I should have done. In my next commit I will make sure to get away with `Eq/Ne` and replace them with `Q.eq/Q.ne`.

krishnavbajoria02 committed:
Fixed incorrect encodings and simplified the code

krishnavbajoria02 committed:
Remove unwanted imports

TiloRC commented:
I looked through the changes and it doesn't look like you've added the tests I've suggested yet. Adding more intelligently chosen tests should help catch potential bugs.

TiloRC commented on sympy/logic/tests/test_euf_solver.py:
```diff
@@ -0,0 +1,1801 @@
+        assert simple_disjunction not in (False, True)
+        z3_sat = z3_satisfiable(simple_disjunction) is not False
+
+        assert sympy_sat == z3_sat
```
it looks like this test is failing becasue sympy sat isn't the right kind of object

krishnavbajoria02 committed:
Covered more cases for explanation of conflicts and added more tests

krishnavbajoria02 committed:
Added alternate dfs path to explore more congruent equations

krishnavbajoria02 committed:
Remove unused parameters

krishnavbajoria02 committed:
adding import z3 before starting test

TiloRC commented on sympy/logic/tests/test_euf_solver.py:
```diff
@@ -0,0 +1,1938 @@
+    enc.data = [[1]]
+
+    solver, conflicts = EUFTheorySolver.from_encoded_cnf(enc, testing_mode=True)
+    # Should convert properly
```
What does this comment mean?

krishnavbajoria02 replied:
This has been done in the latest commit.

TiloRC commented on sympy/logic/tests/test_euf_solver.py:
```diff
@@ -0,0 +1,1938 @@
+    solver.Initialize({Q.eq(a, b)})
+
+    # Test with no current conflict
+    conflict = solver._generate_minimal_conflict()
+    assert isinstance(conflict, set)
```
It's a little counterintuitive that this returns an empty set when there is no conflict. If you added an empty set as a conflict clause to the sat solver it would make everything unsatisfiable. I think it would be more reasonable to raise an exception or rename the function something like `check`.

TiloRC commented on sympy/logic/tests/test_euf_solver.py:
```diff
@@ -0,0 +1,1938 @@
+    conflict = solver._generate_minimal_conflict()
+    assert isinstance(conflict, set)
+
+    # Test with unknown exception type
+    solver._current_exception_type = 'unknown'
+    conflict = solver._generate_minimal_conflict()
+    assert isinstance(conflict, set)
```
What is solver._current_exception_type? Do we need it?

krishnavbajoria02 replied:
So it is basically the two types of exception `EUFEqualityContradiction` Exception and `EUFDisequalityContradiction` Exception classes and when we encounter one of them the `solver._current_exception_type` changes to the appropriate type of exception, whether it is contradicting an equality or a disequality.

TiloRC commented on sympy/assumptions/euf_ask.py:
```diff
@@ -0,0 +1,142 @@
+"""
+EUF Ask Module
+==============
+
+This module implements the interface for querying assumptions using the EUF
+(theory of Equality with Uninterpreted Functions) solver. It follows a scheme
+similar to the LRA SAT-ask interface [see lra_satask.py] and is integrated with a
+DPLL(T) framework.
+
+Key functions:
+    - euf_ask: Main entry point for answering queries using EUF theory.
+    - _is_direct_euf_propagation: Checks for simple propagation cases.
+    - _handle_direct_propagation: Handles direct propagation when a predicate's
+      assumption can be inferred immediately.
+    - check_satisfiability: Determines satisfiability under the EUF theory.
+
+References:
+    [1] DPLL(T): Fast Decision Procedures.
+    [2] Proof-Producing Congruence Closure.
+
+Examples:
+    >>> from sympy.assumptions.euf_ask import euf_ask
+    >>> from sympy.assumptions.ask import Q
+    >>> from sympy import symbols
+    >>> x, y = symbols('x y')
+    >>> euf_ask(Q.prime(x), Q.prime(y) & Q.eq(x, y))
+    True
+
+Doctest:
+    >>> from sympy.assumptions.euf_ask import euf_ask
+    >>> from sympy.assumptions.ask import Q
+    >>> from sympy import symbols
+    >>> x, y = symbols('x y')
+    >>> euf_ask(Q.prime(x), Q.prime(y) & Q.eq(x, y))
+    True
```
You don't need all of these docstrings. They give the false impression that this code is not a temporary stopgap.

TiloRC commented on sympy/assumptions/satask.py:
```diff
@@ -79,16 +79,16 @@ def satask(proposition, assumptions=True, context=global_assumptions,
-    can_be_true = satisfiable(sat_true)
-    can_be_false = satisfiable(sat_false)
+    can_be_true = satisfiable(sat_true, use_euf_theory=use_euf_theory)
+    can_be_false = satisfiable(sat_false, use_euf_theory=use_euf_theory)
```
Rather than having use_euf_theory be a parameter that's passed around everywhere it would probably be better just to define a boolean constant like USE_EUF_THEORY at the top of the file. In other words, avoid changing lines 18, 82 and 85

TiloRC replied:
Ideally satask would always use use_euf_theory. But for now we can disable that since you said it causes some tests to fail.

krishnavbajoria02 commented on sympy/logic/algorithms/euf_theory.py:
```diff
@@ -0,0 +1,278 @@
+        >>> cc.are_equal(a, b)
+        True
+        """
+        self.pending_unions.append((self._flatten(lhs), self._flatten(rhs)))
```
I was thinking not to use flattening much in euf theory solver since flattening is more related to euf congruence closure and I have even defined the flatten function in the congruence closure so it might be confusing, don't you think ? I am open to change it but that was my idea behind this code.

krishnavbajoria02 commented on sympy/logic/tests/test_euf_solver.py:
```diff
@@ -0,0 +1,1938 @@
+        solver.SetTrue(Q.ne(a, b))
+
+
+def test_conflict_generation_equality():
+    """Test conflict generation for equality contradictions."""
+    solver = EUFTheorySolver()
+    solver.Initialize({Q.eq(a, b), Q.ne(a, b)})
+
+    # Set up contradiction scenario
+    solver.SetTrue(Q.eq(a, b))
+
+    # This should generate a conflict appropriately
```
Yes, I'll remove this too in my next commit. Thanks.

krishnavbajoria02 committed:
Removing unnecessary lines of code

- Removed unnecessary print statements
- Removed unnecessary tests that weren't testing anything
- Removed unnecessary docstrings

krishnavbajoria02 committed:
Removed unwanted alias

krishnavbajoria02 committed:
Updated the test to check for the exact length of the conflict clause
